substrate/frame/distribution/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// Re-export all pallet parts, this is needed to properly import the pallet into the runtime.
+pub use pallet::*;
+mod functions;
+mod types;
+pub use types::*;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		/// https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/frame_runtime_types/index.html
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Type to access the Balances Pallet.
+		type NativeBalance: fungible::Inspect<Self::AccountId>
+			+ fungible::Mutate<Self::AccountId>
+			+ fungible::hold::Inspect<Self::AccountId>
+			+ fungible::hold::Mutate<Self::AccountId, Reason = Self::RuntimeHoldReason>
+			+ fungible::freeze::Inspect<Self::AccountId>
+			+ fungible::freeze::Mutate<Self::AccountId>;
+
+		/// Treasury account Id
+		#[pallet::constant]
+		type PotId: Get<PalletId>;
+
+		type RuntimeHoldReason: From<HoldReason>;
+
+		/// This the minimum required buffer time period between project nomination
+		/// and payment/reward_claim from the treasury.
+		#[pallet::constant]
+		type BufferPeriod: Get<BlockNumberFor<Self>>;
+
+		/// Maximum number projects that can be accepted by this pallet
+		#[pallet::constant]
+		type MaxProjects: Get<u32>;
+
+		/// Epoch duration in blocks
+		#[pallet::constant]
+		type EpochDurationBlocks: Get<BlockNumberFor<Self>>;
+	}
+
+	/// A reason for placing a hold on funds.
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		/// Funds are held for a given buffer time before payment
+		#[codec(index = 0)]
+		FundsReserved,
+	}
+
+	/// Number of spendings that have been executed so far.
+	#[pallet::storage]
+	pub(super) type SpendingsCount<T: Config> = StorageValue<_, SpendingIndex, ValueQuery>;
+
+	/// Executed spendings information.
+	#[pallet::storage]
+	pub(super) type CompletedSpendings<T: Config> =
```

Comment:
why do we need to store those?

Reply:
Here I'm being paranoid: If a spend failed to be executed for some reason, it will remain in the Spends storage for future execution. its status however will be updated accordingly, this way we can differentiate failed spends from new/non_completed ones.

Reply:
so this is `FailedSpends` not `CompletedSpends`?

Reply:
Oh ok sorry, I got a bit lost in my explanation, but now I get what you're asking:
This is open to debate and not related to code efficiency, but I just think that due to the "immutable history" aspect of the blockchain, keeping this data (Funded Projects infos in `CompletedSpends`) might have some value down the line.

Reply:
so it is useless atm and in that case we should remove it. we can't just save everything because it may have some value in the future. there are cost associated on every stored values and we need to minimize them.

Reply:
Removed the `CompletedSpends` storage.